73_ArithmeticBench/main.cpp

```diff
@@ -0,0 +1,949 @@
+#include "SimpleWindowedApplication.hpp"
+#include "CEventCallback.hpp"
+#include "nbl/application_templates/MonoAssetManagerAndBuiltinResourceApplication.hpp"
+#include "app_resources/common.hlsl"
+
+using namespace nbl;
+using namespace core;
+using namespace system;
+using namespace asset;
+using namespace ui;
+using namespace video;
+
+// method emulations on the CPU, to verify the results of the GPU methods
+template<class Binop>
+struct emulatedReduction
+{
+	using type_t = typename Binop::type_t;
+
+	static inline void impl(type_t* out, const type_t* in, const uint32_t itemCount)
+	{
+		const type_t red = std::reduce(in,in+itemCount,Binop::identity,Binop());
+		std::fill(out,out+itemCount,red);
+	}
+
+	static inline constexpr const char* name = "reduction";
+};
+template<class Binop>
+struct emulatedScanInclusive
+{
+	using type_t = typename Binop::type_t;
+
+	static inline void impl(type_t* out, const type_t* in, const uint32_t itemCount)
+	{
+		std::inclusive_scan(in,in+itemCount,out,Binop());
+	}
+	static inline constexpr const char* name = "inclusive_scan";
+};
+template<class Binop>
+struct emulatedScanExclusive
+{
+	using type_t = typename Binop::type_t;
+
+	static inline void impl(type_t* out, const type_t* in, const uint32_t itemCount)
+	{
+		std::exclusive_scan(in,in+itemCount,out,Binop::identity,Binop());
+	}
+	static inline constexpr const char* name = "exclusive_scan";
+};
+
+class ArithmeticBenchApp final : public examples::SimpleWindowedApplication, public application_templates::MonoAssetManagerAndBuiltinResourceApplication
```

Comment:
please leave a comment that the whole reason we have a swapchain is because of Nsight being limited